### PR TITLE
nll type annotations in multisegment path

### LINKED ISSUE #54574: nll should respect lifetime annotations from multi-segment paths
This example should not compile: 

```rust
#![feature(nll)]

struct A<'a> { x: &'a u32 }

impl<'a> A<'a> {
    fn new(x: &'a u32) -> Self {
        Self { x }
    }
}

fn foo<'a>() {
    let v = 22;
    let x = A::<'a>::new(&v);
}

fn main() {}
```

But it does =)

### PR DESCRIPTION
**Pull request by nikomatsakis**
This turned out to be sort of tricky. The problem is that if you have a path like

```
<Foo<&'static u32>>::bar
```

and it comes from an impl like `impl<T> Foo<T>` then the self-type the user gave doesn't *directly* map to the substitutions that the impl wants. To handle this, then, we have to preserve not just the "user-given substs" we used to do, but also a "user-given self-ty", which we have to apply later. This PR makes those changes.

It also removes the code from NLL relate-ops that handled canonical variables and moves to use normal inference variables instead. This simplifies a few things and gives us a bit more flexibility (for example, I predict we are going to have to start normalizing at some point, and it would be easy now).

r? @matthewjasper -- you were just touching this code, do you feel comfortable reviewing this?

Fixes #54574

### CODE REVIEW

File: src/librustc/ty/subst.rs [hunk 1]
```diff
@@ -564,3 +537,98 @@ impl<'a, 'gcx, 'tcx> SubstFolder<'a, 'gcx, 'tcx> {
...
+/// example, in this code:
+///
+/// ```
+/// struct Foo<T> { }
+/// impl<A> Foo<Vec<A>> { fn method() { } }
+/// ```
+///
+/// when you then have a path like `<Foo<Vec<&'static u32>>>::method`,
+/// this struct would carry the def-id of the impl along with the
+/// self-type `Vec<u32>`. Then we can instantiate the parameters of
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc/ty/subst.rs]: this is a bit messed up, yeah, I think the `Vec` is adding pretty little, actually.

File: src/librustc/infer/nll_relate/mod.rs [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+
+pub trait TypeRelatingDelegate<'tcx> {
+    /// Push a constraint `sup: sub` -- this constraint must be
+    /// satisfied for the two types to be related. `sub` and `sup` may
+    /// be regions from the type or new variables created through the
+    /// delegate.
+    fn push_outlives(&mut self, sup: ty::Region<'tcx>, sub: ty::Region<'tcx>);
+
+    /// Creates a new universe index. Used when instantiating placeholders.
+    fn next_subuniverse(&mut self) -> ty::UniverseIndex;
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc/infer/nll_relate/mod.rs]: I did but that PR has not landed yet =)

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc/infer/nll_relate/mod.rs]: I guess it will land first though

**Review by matthewjasper — APPROVED** [review]
LGTM. Not sure how confident I am reviewing a PR of this size though.

**Review by pnkfelix — COMMENTED** [review]
If you end up rebasing this PR, you might as well move 24bee005b646692b265af026cf863acf9a322460 down next to e339e84fffb3156aaa78fbf686436de02ba4cf48 and then squash the former into the latter.

(But I haven't seen anything yet that would stop me from r+'ing this, I'm just noting it.)

**Review by pnkfelix — COMMENTED** [review]
hooray for the new `UserTypeAnnotation` enum! I hope this will make the corresponding visit  methods easier to grok

**Review by pnkfelix — COMMENTED** [review]
Eeek regarding e7ed997d87a1769bd68a86412ae2a417e4d36b99; its a shame we couldn't continue keeping the contents of `nll_relate` hidden away under `rustc_mir::borrow_check::nll` ...

File: src/librustc/ty/subst.rs [hunk 3]
```diff
@@ -564,3 +537,98 @@ impl<'a, 'gcx, 'tcx> SubstFolder<'a, 'gcx, 'tcx> {
...
+
+/// Stores the user-given substs to reach some fully qualified path
+/// (e.g., `<T>::Item` or `<T as Trait>::Item`).
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
+pub struct UserSubsts<'tcx> {
+    /// The substitutions for the item as given by the user.
+    pub substs: &'tcx Substs<'tcx>,
+
+    /// The self-type, in the case of a `<T>::Item` path (when applied
+    /// to an inherent impl). See `UserSubsts` below.
```
Comment (pnkfelix) [inline src/librustc/ty/subst.rs]: I don't understand where the reader is meant to be led by the "See `UserSubsts` below."

Comment (pnkfelix) [reply src/librustc/ty/subst.rs]: (Probably a typo for `UserSelfTy`.)

File: src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#![feature(nll)]
+
+// Check that substitutions given on the self type (here, `A`) carry
+// through to NLL.
+
+struct A<'a> { x: &'a u32 }
+
+impl<'a> A<'a> {
+    fn new<'b, T>(x: &'a u32, y: T) -> Self {
```
Comment (pnkfelix) [inline src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs]: why is there a `'b` here?

Comment (pnkfelix) [reply src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs]: (I see the same thing occuring in the other tests. To be honest I had thought that we didn't allow unreferenced lifetimes in generic binders ... but I guess that is just a lint, when it comes to lifetimes...?)

File: src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        Self { x }
+    }
+}
+
+fn foo<'a>() {
+    let v = 22;
+    let x = A::<'a>::new(&v, 22);
```
Comment (pnkfelix) [inline src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs]: I think you and I mirrored each other on our respective PR's; there should be an `//~ ERROR` on this line, right?

**Comment by pnkfelix** [comment]
I had a slew of nits. But none of them should block landing this PR, and frankly, I want to rebase my own work pronto on top of this.

File: src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs (hunk 4 above)
Comment (nikomatsakis) [reply src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs]: copy and paste error, I think

File: src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs (hunk 5 above)
Comment (nikomatsakis) [reply src/test/ui/nll/user-annotations/method-ufcs-inherent-1.rs]: Eep yes, will add.

File: src/librustc/ty/subst.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc/ty/subst.rs]: yes, fixing

**Comment by nikomatsakis** [comment]
@pnkfelix 

> Eeek regarding e7ed997; its a shame we couldn't continue keeping the contents of nll_relate hidden away under rustc_mir::borrow_check::nll ...

Well, I was hoping they will get unified at some point, in which case it makes sense.

**Pull request merged**